### PR TITLE
Run bars fix

### PR DESCRIPTION
**Pull request by proskurin**
# Description

Fixes # (issue)

Run bars generation issue. Now the code algorithm looks more stable and similar to the one described in a book

- [x] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] test_run_data_structures.py

**Test Configuration**:
* Ubuntu 16.04
* Pycharm CE

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: mlfinlab/data_structures/run_data_structures.py [hunk 1]
```diff
@@ -90,45 +93,55 @@ def _extract_bars(self, data):
...
             elif imbalance < 0:
                 self.imbalance_array['sell'].append(abs(imbalance))
-                # Set zero to keep buy and sell arrays synced
-                self.imbalance_array['buy'].append(0)
                 cum_theta_sell += abs(imbalance)
 
-            if not list_bars and np.isnan(self.exp_buy_proportion):
-                self.exp_buy_proportion, self.exp_sell_proportion = self._get_expected_imbalance(
-                    self.exp_num_ticks, self.imbalance_array)
+            if not list_bars and np.isnan([self.exp_imbalance['buy'], self.exp_imbalance['sell']]).any():
```
Comment (Jackal08) [inline mlfinlab/data_structures/run_data_structures.py]: Can we add a comment to explain what is happening here? 

* Can we assign this bool to a named variable? np.isnan([self.exp_imbalance['buy'], self.exp_imbalance['sell']]).any()

Comment (Jackal08) [reply mlfinlab/data_structures/run_data_structures.py]: Not a big deal. You decide.

File: mlfinlab/data_structures/run_data_structures.py [hunk 2]
```diff
@@ -90,45 +93,55 @@ def _extract_bars(self, data):
...
+                                                                          self.imbalance_array['sell'])
+                if bool(np.isnan([self.exp_imbalance['buy'], self.exp_imbalance['sell']]).any()) is False:
+                    self.exp_buy_ticks_proportion = buy_ticks / cum_ticks
+                    cum_theta_buy, cum_theta_sell = 0, 0  # reset theta after warm-up period
+                    self.warm_up = False
 
             # Update cache
             self._update_cache(date_time, price, low_price, high_price, cum_theta_sell, cum_theta_buy,
-                               cum_ticks, cum_volume)
+                               cum_ticks, buy_ticks, cum_volume)
```
Comment (Jackal08) [inline mlfinlab/data_structures/run_data_structures.py]: Are we missing sell_ticks?

File: mlfinlab/tests/test_run_data_structures.py [hunk 3]
```diff
@@ -105,7 +105,7 @@ def test_run_tick_bars(self):
         """
         Tests the imbalance tick bars implementation.
```
Comment (Jackal08) [inline mlfinlab/tests/test_run_data_structures.py]: Should this rather be run bars? it mentions imbalance. Check this in the other doc strings as well.

File: mlfinlab/tests/test_run_data_structures.py [hunk 4]
```diff
@@ -67,7 +67,7 @@ def test_run_volume_bars(self):
         """
         Tests the imbalance volume bars implementation.
```
Comment (Jackal08) [inline mlfinlab/tests/test_run_data_structures.py]: Mention of imbalance. Should be runs

File: mlfinlab/tests/test_run_data_structures.py [hunk 5]
```diff
@@ -29,7 +29,7 @@ def test_run_dollar_bars(self):
         """
         Tests the imbalance dollar bars implementation.
```
Comment (Jackal08) [inline mlfinlab/tests/test_run_data_structures.py]: Imbalance / Runs

File: mlfinlab/data_structures/run_data_structures.py [hunk 6]
```diff
@@ -171,18 +185,19 @@ def _update_cache(self, date_time, price, low_price, high_price, cum_theta_sell,
         :param cum_ticks: Cumulative number of ticks
         """
         cache_data = self.cache_tuple(date_time=date_time, price=price, high=high_price, low=low_price,
-                                      cum_ticks=cum_ticks, cum_volume=cum_volume, cum_theta_buy=cum_theta_buy,
+                                      cum_ticks=cum_ticks, buy_ticks=buy_ticks, cum_volume=cum_volume,
+                                      cum_theta_buy=cum_theta_buy,
                                       cum_theta_sell=cum_theta_sell)
         self.cache.append(cache_data)
 
     def _get_expected_imbalance(self, window, imbalance_array):
```
Comment (Jackal08) [inline mlfinlab/data_structures/run_data_structures.py]: get expected imbalance or get expected runs?

File: mlfinlab/data_structures/run_data_structures.py (hunk 2 above)
Comment (proskurin) [reply mlfinlab/data_structures/run_data_structures.py]: sell_ticks = cum_ticks  - buy_ticks

File: mlfinlab/data_structures/run_data_structures.py (hunk 6 above)
Comment (proskurin) [reply mlfinlab/data_structures/run_data_structures.py]: Actually, we get expected imbalance. In run bars we calculate the same imbalance, the only difference is that we split buy and sell imbalance

**Pull request merged**